### PR TITLE
ENH: support pipeline extensions in pipeline config

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
This PR adds support of pipeline_extensions in pipelines configuration file with the following restriction:

1. when duplicate extensions are configured in data-prepper-config and pipeline-config, the definition in pipeline_extensions wins
2. pipeline-config is restricted to a single file when configured with pipeline_extensions.
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dlvenable** [comment]
> when duplicate extensions are configured in data-prepper-config and pipeline-config, the definition in pipeline_extensions wins

I would not say this is a restriction. This is the design. Using this here is stating that we want it to be the override.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PipelinesDataFlowModel.java [hunk 1]
```diff
@@ -28,6 +30,11 @@ public class PipelinesDataFlowModel {
     @JsonInclude(JsonInclude.Include.NON_NULL)
     private DataPrepperVersion version;
 
+    @JsonProperty("pipeline_extensions")
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PipelinesDataFlowModel.java:33]: In #2590 this is proposed as `pipeline_configurations`. I tend to think that we should keep this original name.

There are a couple reasons:

1. Extensions are intended for extending Data Prepper. Thus, extensions may need to run before parsing a pipeline configuration.
2. The available options in `pipeline_configurations` should be somewhat different from the `extensions`. Not everything in extensions should be configurable in the pipeline.

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+@Named
+public class ExtensionPluginConfigurationResolver {
+    private final Map<String, Object> extensionMap;
+
+    @Inject
+    public ExtensionPluginConfigurationResolver(final DataPrepperConfiguration dataPrepperConfiguration,
```
Comment (dlvenable) [inline data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java:17]: As noted in another comment, we should not tie extensions configurations directly with pipeline configurations. Can we decouple these somewhat?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java (hunk 2 above)
Comment (chenqi0805) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java:17]: Would you mind if I do it in the other PR: https://github.com/opensearch-project/data-prepper/pull/3340. There has been some tweaks on the interface

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PipelinesDataFlowModel.java (hunk 1 above)
Comment (chenqi0805) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/configuration/PipelinesDataFlowModel.java:33]: Will rename this part

File: data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java (hunk 2 above)
Comment (dlvenable) [reply data-prepper-core/src/main/java/org/opensearch/dataprepper/plugin/ExtensionPluginConfigurationResolver.java:17]: That seems reasonable.

**Review by dlvenable — APPROVED** [review]

**Pull request merged**